rabbitmq/retryable_producer.go

```diff
@@ -0,0 +1,169 @@
+	producer *Producer
+}
+
+type RetryableProducerConfig struct {
+	MaxRetryAttempts int
+	// HealthCheckFactor is a number representing how much N multiplied by backoffConfig.Max time is needed
+	// for a block of code to run w/o returning an error, to consider it healthy.
+	// E.g backConfig.Max = 1min, healthCheckFactor = 2, means that code needs to run 2min at least to be healthy
+	// and retried again starting from backoffConfig.Base the next time it has an error.
+	HealthCheckFactor  int
+	BackoffConfig      *backoff.Config
+	RabbitClientConfig *ClientConfig
+}
+
+func NewRetryableProducer(
+	newClientFactory func(ctx context.Context, config *ClientConfig) (RabbitMQClientInterface, error),
+	config RetryableProducerConfig,
+	logger logger.StructuredLogger,
+	metric Metric,
+) *RetryableProducer {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	retryableProducer := &RetryableProducer{
+		logger:        logger,
+		metric:        metric,
+		clientFactory: newClientFactory,
+		config:        config,
+		cancel:        cancel,
+	}
+
+	go retryableProducer.initProducer(ctx)
+
+	return retryableProducer
+}
+
+func (p *RetryableProducer) Publish(
+	exchange,
+	key string,
+	mandatory,
+	immediate bool,
+	expiration string,
+	body []byte,
+	args amqp.Table,
+) error {
+	p.mu.RLock()
+	producer := p.producer
+	p.mu.RUnlock()
+
+	if producer == nil {
+		return stacktrace.NewError("RabbitMQ Producer client not connected")
+	}
+
+	err := producer.Publish(exchange, key, mandatory, immediate, expiration, body, args)
+	if err != nil {
+		return stacktrace.Propagate(err, "failed to publish RMQ message")
+	}
+
+	return nil
+}
+
+func (p *RetryableProducer) newProducer(ctx context.Context) (*Producer, error) {
+	if ctx.Err() != nil {
+		p.logger.Info("received context cancel")
+		return nil, nil
+	}
+
+	client, err := p.clientFactory(ctx, p.config.RabbitClientConfig)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "RabbitMQ Failed to init client")
+	}
+
+	producer, err := NewProducer(client, p.logger, p.metric)
+	if err != nil {
+		connCloseErr := client.Close()
+		p.logger.Error("cannot close RabbitMQ client connection", zap.Error(connCloseErr))
+		return nil, stacktrace.Propagate(err, "RabbitMQ Failed to create new producer")
+	}
+
+	return producer, nil
+}
+
+func (p *RetryableProducer) initProducer(ctx context.Context) {
+	producerBackoff := backoff.NewBackoff(p.config.BackoffConfig)
+	currentRetryAttempts := 0
+
+	for {
+		producer, err := p.newProducer(ctx)
+		if err != nil {
+			p.logger.Error("producer connection failed with error", zap.Error(err))
+
+			if p.config.MaxRetryAttempts != 0 && currentRetryAttempts > p.config.MaxRetryAttempts {
+				p.logger.Info("retry attempts exceeded")
+				return
+			}
+
+			backoffDuration := producerBackoff.Next()
+
+			select {
+			case <-ctx.Done():
+				p.logger.Info("received context cancel")
+				return
+			case <-time.After(backoffDuration):
+				continue
+			}
+		}
```

Comment:
This must be in a separate for loop that is making the backoff retry on creating a producer - can be moved to a separate method for example `newProducerWithBackoff()` .

This function will use a new backoff on every invocation, the following lines will be moved there too.

```
	producerBackoff := backoff.NewBackoff(p.config.BackoffConfig)
	currentRetryAttempts := 0
```	

So the main for loop in `initProducer` will be 
```
for {
  // create producer (retry until a producer is successfully created)
  // set producer
  // wait for the producer to be closed
}

Reply:
Yeah looks better now. Pushed the changes.